(Moved to #1351 !) Implemented Quantum Ranks optional feature

AllUniversal opened the pull request:
+Title

I gave implementing quantum ranks a try and am quite happy with the result.

Currently this supports all hand detection stuff (`X same` and `Straight`), I don't know how inefficient it gets but for vanilla hands it should be fine. (I tested 5 cards + 13 ranks per card and didn't notice any lag, and 7 vanilla cards with no lag either. Will do further testing if needed.)
`Shortcut` and `Four Fingers` are also supported.
Edge cases may still come up but some are already accounted for. (Like dealing with `rank.straight_edge = true` properly or forking paths in `.next`/`.prev` of a rank)

If you use the added `Card:is_rank()` function, this also supports any Jokers/Enhancements/Etcs. changing a card's rank, because I added a new `context` for getting a card's ranks (`context.get_ranks`). 
-> Importantly, I have not (and probably will not) implement cards holding more than one rank _independently of any effects_. As in: If you want specific cards to have more than one rank, with this implementation you'd have to make a custom `Enhancement` or similar, which returns the intended ranks during `context.get_ranks` evaluation for the card it's applied to.

I tried commenting my code but further documentation will follow (like a PR to the Wiki repo).

Patching the vanilla Jokers could use some attention by a regexpert (<- not me!).

Any feedback is welcome! I fully expect my straight algorithm (especially) to be replaced by something better eventually, but for now I just wanted to get something working. :)


## Additional Info:
<!-- Don't worry too much if you don't know what these are or how to fill them. It's mostly reminders for maintainers ;) -->
- [ ] I didn't modify api's or I've made a PR to the [wiki repo](https://github.com/Steamodded/wiki).
- [x] I didn't modify api's or I've updated lsp definitions.
- [x] I didn't make new lovely files or all new lovely files have appropriate priority.

AllUniversal committed:
Quantum Ranks: Straight detection implemented and Card:get_ranks() + more

+Title, this is working with Shortcut and Four Fingers.
-> Some edge cases have already come up (like straight_edge ranks), I've fixed the ones I've found but I'm certain there will be more.

!Ironically the only parts that kept breaking stuff by the end were ""optimizations"" I thought of... maybe I should stop doing that woops.

+Card:get_ranks() and Card:is_rank() functions
-> These come with a new context type: context.get_ranks
-> It only has the parameter "other_card" for the playing card that its called from.
-> Returning "ranks" equal to a table of "SMODS.Rank"s will override the card's rank.

!All of these features require SMODS.optional_features.quantum_ranks to be true

!Documentation is lacking and there's still much to do, more commits to follow. Eventually. :)

AllUniversal committed:
Replaced card_rep.straight_pos with a set of all used card_reps + some optimizations

*Title, straight_pos somehow resulted in card_reps being used multiple times which greatly hurt performance and didn't return the right cards.
-> The new system seems to behave correctly and also removed any immediate performance problems with a crude wild rank implementation (-> simply returning all SMODS.Ranks in context.get_ranks)

+A new exit condition for the straight_calculation based on the amount of card_reps checked or the total size of the best_straight so far.
-> These should be edge case free (in theory) (hopefully) (please?)

*context.get_ranks' "ranks" field now has the card's rank set as default.

AllUniversal committed:
Updated/Implemented get_X_same(), added QoL functions, fully patched vanilla "get_id() == " and fixed some bugs

+get_X_same() now works with quantum ranks too!
-> This also means the previously injected "or_more" parameter is now part of the "overrides.lua" function definition and the according patches have been removed from "playing_cards.toml".
+Added Card:is_any_rank() and SMODS.get_rank_by_id() functions.
-> is_any_rank() allows passing a table of ranks, making Hack/Fibonacci/etc. easier to patch/implement.
-> get_rank_by_id() does what its name implies.
!I considered adding a lookup table alike SMODS.Ranks but by IDs, but for now I'll leave it.
*/+Most vanilla jokers now use Card:is_rank() instead of get_id()
-> Exceptions are Even Steven, Odd Todd and Cloud Nine, due to implementation and potential performance problems.
+Added some more comments to get_straight()
-Removed semi functional implementation of 1-squashing optimization for get_straight().
-> I'll see if I'll get back to it, but it seems to run fine without.

AllUniversal committed:
Merge branch 'main' into Quantum-Ranks

AllUniversal committed:
Added some LSP defs and fixed a mistake in straight calculation

+LSP defs for SMODS.get_rank_by_id() and get_ranks context field.
*rank.straight_edge edge case handling had a mistake (probably) where the used_c_reps where being set incorrectly for the second straight evaluation.

AllUniversal committed:
Important straight calc optimization

*c_reps which have appeared in a ret_straight are skipped as starting points.
-> As long as this doesn't cause any edge cases this may be a crucial optimization, because instead of checking every card_rep, it now only checks every branch of possible straights once.
-> Please work :) (I beggeth, I pleadeth)

AllUniversal committed:
Added ´source_context´ parameter to ´context.get_ranks´ and fixed small oversight

+´get_ranks´ context now has a ´source_context´ field, allowing more granular conditional rank changing (like, depending on what joker may have called is_rank(), though I'm unsure if that would also require a ´source_obj´ param)
*Fixed missing ´pairs()´ in ´Card:is_any_rank()´

AllUniversal commented:
I realized that get_X_same will actually need another look, as right now cards can be counted multiple times, which at the very least is inconsistent with how they work in straights. I'll try fixing it.

AllUniversal committed:
Fixed `get_X_same()` behaving incorrectly for Full House calculation.

*Title, this required all my brainpower.
->The issue is now fixed and Full House can be correctly evaluated in the following scenario:
1. Jacks are also considered Kings and Queens.
2. The selected hand is K K Q Q J, or K Q J, or Q J J, or K Q Q J J, or K Q J J,
evaluating as Full House, Pair, Three of a kind, Four of a kind and Three of a Kind respectively.
! The last hand is important; Technically there's also a Two Pair which my implementation does NOT pick up, but (barring any edgecases I'm unaware of), this will never be relevant because there will always be a Three of a kind (or better) instead.
*Also fixed another missing `pairs()` in `utils.lua`

AllUniversal committed:
Merge branch 'main' into Quantum-Ranks

AllUniversal commented:
Alright, the above issue has been fixed and `get_X_same()` now works just like straight calculation. 
This required changing how Full House and Flush House check their respective `PokerHandParts` and also how those call `get_X_same()`, specifically, only `_5` and `_all_pairs` now call the function with `or_more=True`. (`_2`, `_3` and `_4` no longer do)

AllUniversal committed:
Removed unintentionally committed changes from other PR (Woopsie)

-Title

AllUniversal committed:
Merge branch 'Quantum-Ranks' of https://github.com/AllUniversal/smods into Quantum-Ranks

AllUniversal committed:
Added `eval_getting_ranks` default `source_context` for `get_ranks()` during hand type evaluation + `Card:is_any_rank()` fix

+Title. this default `source_context` allows you to selectively have cards only affect / not affect hand type evaluation.
*/+`Card:is_any_rank()` now works with rank ids or keys too.

AllUniversal commented:
Alright, I tested 5 cards with each card counting as its own rank and its .next rank (picture below). No lag yet on my end (really shouldn't be lagging yet).

<img width="2035" height="1133" alt="image" src="https://github.com/user-attachments/assets/00e921ce-9728-4708-93a2-5ca411eb8357" />
<img width="2033" height="1139" alt="image" src="https://github.com/user-attachments/assets/8da1fdb9-2d97-4e3d-b1ad-e3e176289c51" />

Straight Flush with 7 cards: 

<img width="2033" height="1139" alt="image" src="https://github.com/user-attachments/assets/8963c852-151b-44b8-a279-550059a89cbf" />

Flush House with 8 cards: (2 Jacks, 10->Jack, 9, 8-> 9 => J J J 9 9

<img width="2036" height="1138" alt="image" src="https://github.com/user-attachments/assets/9272a6ce-f265-439c-9b66-9a2395d4d201" />

Performance wise it seems fine, but I don't know why the second to last hand (straight flush) doesn't evaluate as a Flush House.. 
Will look into it.

AllUniversal commented:
Alright, `Full House` calculation might need its own function. I'll give it some more thought first though.

nh6574 commented:
Nice work on this!! I'm excited to see this merged

Though I would recommend turning it into a draft PR if you're still working on it

AllUniversal committed:
Reworked `get_X_same()` and the corresponding hand parts and added dedicated `Full House` / `Two Pair` functions.

*/+Title, this was necessary because `Full House` and `Two Pair` didn't work.

-> This means `get_X_same()` is now once again always called with `or_more=true`, and no longer accounts for quantum cards appearing multiple times, instead it returns all possible 'X (or >X) same's.
-> `Full House` and `Two Pair` now call `get_quantum_full_house(_3, _2)` and `get_quantum_two_pair(_2)` respectively, these functions then check whether an unoverlapping `Pair` and `Three of a Kind` or a set of pairs exists.

!This was way harder than straight evaluation lol. (I may have overcomplicated it here and there, this solution is super simple though. I hope it works.)

AllUniversal committed:
Removed debug print statement

-Woopsie

AllUniversal committed:
Added `Stone card` compatibility

+Title

AllUniversal commented:
Realized I need to implement some more stuff, updated above message.

AllUniversal committed:
Moved `Card:is_rank()`, `Card:get_ranks()`, etc. into `utils.lua`, added `Card:is_face()` compatibility (in `overrides.lua`) and added `no_mod` argument / flag to `context.get_ranks`

*Title, those functions didn't need to be part of the `card_is_rank.toml` lovely patch.
+Title, `Card:is_face()` now works with quantum ranks and uses a new source_context, `source_context.is_face_getting_ranks`, to allow exclusively returning certain ranks during / outside `is_face()` evaluation.
+Title, `context.get_ranks` now takes a new return flag, `no_mod`, to ignore future returned `ranks`.

AllUniversal committed:
Fixed `Card:get_ranks()` returning a rank for stone cards if quantum_ranks are toggled off.

*Title

AllUniversal committed:
Reordered `Card:is_face()` Pareidolia check and `get_ranks()` call.

*Title, to avoid redundant context calculation.

AllUniversal committed:
Removed faulty empty check before returning in `Card:get_ranks()`

-Title, this now allows overriding a card's ranks to be no rank at all.

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Improved functionality of `Rank.straight_edge`

*Title, it's now `Rank.straight_edges`, a map of ranks from which going into this rank will terminate a straight.
-> I added this because I took a look at Paperback's Apostle implementation / definition and realised that its behavious wasn't currently supported that well.
In theory it should now be possible to just not add the Apostle to the Ace's `straight_edges`, which will mean the straight calculation starting from the Apostle will pass through the Ace instead of terminating at it.

! I think the vanilla implementation (which still gets run if `SMODS.optional_features.quantum_ranks = false`) might need to be updated.

AllUniversal committed:
Updated vanilla `get_straight()` implementation to work with updated `Rank.straight_edges`

*Title, this means vanilla straight calc now also allows Paperback-Apostle-like straight rank shenanigans. (In theory)

AllUniversal committed:
Fixed (potential) edge case when checking `#best_straight >= #hand - 1`

*Title, also added `#best_straight >= #hand` to the `recursive_get_straight()` function.

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal commented:
I tested a simple implementation of Ortalab's "Scenic Route" Joker (which allows Straights to contain pairs),
should be as simple as this;

<img width="2048" height="1177" alt="image" src="https://github.com/user-attachments/assets/98cafc56-aa44-4b51-854d-8935cbdcae20" />

<img width="1243" height="644" alt="image" src="https://github.com/user-attachments/assets/bd614c88-3555-476d-91e8-e9cfa9a6337a" />

AllUniversal committed:
Added new flag to toggle optional optimization, `quantum_straight_min_return`

+Title, this was an optimization I considered but had discarded due to potentially returning non-longest straights.
-> If you want 20 card hands and quantum ranks you might have to toggle this on.
!Check the PR for further info.

AllUniversal committed:
Removed `context` argument from `is_rank()` calls.

-Title, instead `is_rank()` etc. now take a `flags` param (used by eval
 for `eval_getting_ranks`)

-> I hope the `SMODS.context_stack` PR gets merged, else I'll lose functionality here.

AllUniversal committed:
LSP defs mini update

*Title

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Fixed `flags` param of `get_ranks()` etc. not being set in all cases and added rank de-duplication.

*/+Title

AllUniversal committed:
Minor LSP defs update

*Title

AllUniversal committed:
Changed `Card:get_ranks()` etc. to use `SMODS.has_no_rank()`

*Title

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Implemented `Wild Rank` API

*Title. """ready for review""" my ass

AllUniversal committed:
Updated `SMODS.has_no_rank()` and fixed a minor Wild rank straight calc bug

*Title

AllUniversal commented:
I've once again realized that the current `straight_edges` implementation causes some unlikely edge cases. I don't know if I'll fix them though, because they only occur with custom ranks, in very specific (and as mentioned, unlikely) scenarios. 
In case this gets merged, if such issues are encountered frequently I'll look into it again, but until then I can't get myself to redo straight calc again. :')

AllUniversal committed:
Added `Card:is_parity()` function

+Title, thanks nh6574 for the idea :)

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge branch 'main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Fixed vanilla `Four of a kind` also counting as `Two Pair` and Joker retriggers on `context.get_ranks`

*Title

Aurelius7309 commented on src/overrides.lua:
```diff
@@ -593,74 +593,457 @@ G.FUNCS.change_stake = function(args)
 end
 
+
+function get_quantum_full_house(_3, _2)
```
This and `get_quantum_two_pair` are too specific and require a lot of extra effort for hand types that combine multiple sets of some amount of cards each (think a fuller house 4+2, two triple, three pair, etc.). There needs to be a utility that can do this in the general case. It'd accept a combination in form of a table like `{length:amount}` and the corresponding hand parts indexed by their length. I can give implementing this a shot myself if you don't want to.

Aurelius7309 commented on src/utils.lua:
```diff
@@ -2684,6 +2709,116 @@ function SMODS.is_eternal(card, trigger)
+    return ret
+end
+
+function Card:is_parity(parity, exclude_faces)
```
Again, face cards shouldn't be special in regards to parity - they inherently don't have one. So there should be no need to explicitly exclude them here

Aurelius7309 commented on src/overrides.lua:
```diff
@@ -593,74 +593,457 @@ G.FUNCS.change_stake = function(args)
+									return ret_straight
+								end
+							end
+							if SMODS.optional_features.quantum_straight_min_return and #ret_straight > min_length then return ret_straight end -- Greatly decreases calculation time but doesn't ensure that the longest straight is found.
```
This optional feature is terrible in terms of API design. If this optimization is at all to be left up to individual mods, then what a mod may require is for it to not be used because it's important for its mechanics that the longest available straight is always detected correctly. This means the optimization should be opt-in instead of opt-out - otherwise such a mod will have no way of ensuring longest straights being found. Even then it might be more of a decision for the player to make instead of for specific mods to mandate?

AllUniversal replied:
Leaving it up to the player may make sense yeah. I don't like having it be a whole other `optional_features` flag anyway, that said it _is_ opt-in, right? You have to set it to true like the other optional features.
It is a little ugly either way. Problem is that I personally can't implement this in a way that allows 30+ card hands without using this optimization. If someone else wishes to take another look, be my guest :)
(However, for _most_ scenarios, the longest straight should be found either way) (I think. Has to be tested before I'm 100% sure)

Aurelius7309 replied:
Whoops! I'm getting tired I guess. What I meant to say is that the optimization should be opt-out instead of being opt-in if it's left up to mods, so that any mod can decide it wants the optimization to be off, as the mods that want to leave it on likely won't have any issues other than performance with it off

Aurelius7309 replied:
No performance related ones, anyway. If there's other performance settings that could be added, this could go along with them in a new tab/page. Seems fine for now as it's the easiest thing to change

Aurelius7309 commented on lsp_def/classes/rank.lua:
```diff
@@ -17,7 +17,8 @@
 ---@field strength_effect? table|{fixed?: number, random?: boolean, ignore?: boolean} Determines how cards with this rank behave when Strength is used. 
----@field straight_edge? boolean Sets if this rank behaves like an Ace for straights. 
+---@field straight_edges? table<Ranks|string, boolean> Sets the rank keys from which going into this rank should terminate a straight. (Vanilla only set for Ace => {'King' = true, '2' = true})
+---@field parity? number The parity of the rank, 0 for Even, 1 for odd, anything else for neither
```
"Anything else" is setting a bad convention, ranks with no parity should leave this blank

Aurelius7309 reviewed (changes requested):
Had a first closer look at this, here's what I've noticed without digging too deep into the technical side of it

AllUniversal committed:
Updated parity according to feedback

*Title

AllUniversal commented on src/overrides.lua:
```diff
@@ -593,74 +593,457 @@ G.FUNCS.change_stake = function(args)
+
+		-- TODO: Optimization that may or may not be worth it: 
+		-- Remove all card_reps that share a rank with a 1-rank card_rep at that rank. Repeat until none are removed.
+		-- Also, collapse any N card_reps that share N ranks into one of those ranks each. (Hard to implement / too inefficient maybe)
```
Yeah, doing it for all card_reps / all ranks probably would be too much. I kinda hate this optimization :) Because straight calculation isn't perfect anyway probably. So replacing it if it performs badly seems better anyway. That said, I first gotta know how badly it performs.

AllUniversal committed:
Implemented general `get_quantum_pairing()` function as requested

+Title, this should allow custom hands like a 4+2 extended Full House more easily, and (in theory) also supports Three Pair, etc. (but may perform badly then)

Aurelius7309 commented on src/overrides.lua:
```diff
@@ -593,74 +593,486 @@ G.FUNCS.change_stake = function(args)
+	local used_pairs = {}
+	if base then
+		for _, pairs in ipairs(base) do
+			for pair, over_by in pairs(pairs) do
```
This is cursed, please don't call a local variable pairs

cg-223 replied:
pairs(pairs)

AllUniversal replied:
Woopsie

AllUniversal committed:
Woopsie

*We love overshadowing builtins

AllUniversal commented:
@Aurelius7309 Of course! 
The way I implemented it works like you mentioned; the returned `ranks` flag overrides the `context.ranks`, to add ranks you should be able to do what you did above or similar, so I gotta see why that isn't working...
Having multiple calculation targets add ranks should work fine as long as each is implemented correctly, right? `context.ranks` gets updated with the returned flags between calculations. (In `utils.lua:~1755`).

Importantly; The way I did this, Quantum Rank effects are inherently order dependent, so imagine this;
Joker A:    7s also count as 2s
Joker B:    2s also count as 4s 
If Joker A is eval'd before Joker B, played 7s have the ranks 7, 2 and 4.
If Joker B is eval'd before Joker A, played 7s have the ranks 7 and 2.
This is true depending on how Joker B checks whether the card is a 2;
True, if Joker B uses `context.ranks` as reference
False, if Joker B uses `Card:get_id()`
In the latter case, Joker B will only ever add `4` as a Rank if the played card is a vanilla `2`.

I can't think of a way of doing this that also allows removing/setting ranks and isn't ordered, but tell me if you have any suggestions.

I'll have to check why `unpack(context.ranks)` isn't working as I expect here.

Also, I have realized that there's a problem with infinite context loops (I believe you mentioned that on Discord).
When a Quantum Rank Joker calls `SMODS.has_enhancement()` while there's a Quantum Enhancement Joker that calls `Card:is_rank()`, the contexts call each other. (The Jokers/Calc Targets technically don't even have to be separate).
I have thought of a way to (hopefully) fix this and I'll make a separate PR soon™️.

AllUniversal commented:
You right you right, I'll change it to be a map instead of a list.
And I'll see about the conversion breaking :) Very strange.

AllUniversal committed:
Changed `context.ranks` to be a map instead of a list.

*Title, this makes checking for other quantum ranks a lot easier.

!Card:get_ranks() still returns a list though.

AllUniversal committed:
Redid `context.get_ranks` to not be bad

*Title, now accepts two return flags, `ranks_fixed` and `ranks_changed`, for setting or adding/removing ranks respectively.
-> Before you would have had to copy the `context.ranks` and return it + your ranks to not
 accidentally change the `context.ranks` in the wrong places.

Aurelius7309 commented:
Thanks, I'll have another go at testing as well. I'd be in favor of changing Card:get_ranks() returns as well for consistency, are there any reasons against this?
#875: will do :)

AllUniversal commented:
I've decided to redo the `get_ranks` context to use two accepted return values instead of just `ranks = {...}`.
It now accepts `ranks_fixed` and `ranks_changed`. Examples;
Returning `ranks_fixed = {['5'] = true}` would make all cards count as only 5s.
Returning `ranks_changed = {['5'] = true}` would make all cards additionally count as 5s.
Returning `ranks_changed = {['5'] = false}` would make all cards NOT count as 5s. (Assuming later calculation targets don't override this)

AllUniversal committed:
Changed `Card:get_ranks()` to return a map and added missing Wild Rank support to `Card:is_parity()`

*/+Title

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Added `get_ranks` to `SMODS.is_getter_context()`

+Title

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Added some crash safeguards to rank functions

+Title, if called on non-playing cards

AllUniversal committed:
Added `SMODS.get_rank_tally()`

+Title, useful for effects that need to know whether X ranks are in a list of Cards.

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Missing `end`

+Title

AllUniversal committed:
Renamed `Card:is_any_rank()` to `Card:is_ranks()` and added an `all` argument

*Renaming it to avoid confusion with `SMODS.has_any_rank()`
+The `all` param allows checking whether a card is `all = true` all ranks or `all = false` any rank passed to the function.

AllUniversal committed:
Updated `Card:is_ranks()` to avoid Wild Rank cards counting as more than one rank + Added LSP defs to `Card` rank functions

*Title, they now may count as one of the ranks passed to `Card:is_ranks()`
+Also added a safety conversion in case the function gets called with a single rank instead of a list of ranks.
+Title, lsp defs for `Card:is_rank()` etc.

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Replaced `straight_edges` with new `VirtualRank` class

-/*/+Title, checkout the PR (#838) for details

AllUniversal commented:
The recent commit contains some breaking changes; I've replaced `straight_edges` entirely, in favour of a new class, `SMODS.VirtualRank`. It allows for creating rank "redirects" specifically for the purpose of straight calculation. 
By default this includes the `VirtualRanks` `Ace_high` and `Ace_low`, which appropriately redirect the `King` and `2`'s `.next`/`.prev`s and replace the `Ace` for straight calc.

To make this easier **I've changed `.next` and `.prev` to be maps instead of lists**, this means that;
A. their definitions have to be adjusted for custom ranks
and B. it is now recommended that each Rank defines its `strength_effect` explicitly, with the new `fixed_next` and `fixed_prev` fields set to the appropriate rank keys.

Though `VirtualRanks` add quite a bit of complexity, they also avoid many of the problems the simpler `straight_edges` implementation had. (I think (: )

AllUniversal committed:
More LSP defs and small fix

+Title
*Title, VirtualRank's get_straight_next() now returns next_wrap additionally instead of only.

AllUniversal committed:
Updated `SMODS.has_no_rank()` to correctly account for wild ranks + small fix

*Title

AllUniversal committed:
Reworked `get_straight()` because I realized it had a problem

*Title, see PR (#838) for more info.

AllUniversal committed:
Adjusted `get_straight()` optimizations further

*Title (PLEASEEEE WOOOORKK)

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Added `is_royal` field + functions for `SMODS.Rank` and `Card`

+Title, as a simple semi-fix for Royal Flush detection with Quantum Ranks.

! This does mean that a King-High Straight Flush with a Nine that also counts as a 10 will be considered a Royal Flush. But it's just cosmetic anyway, so I guess it's fine.

AllUniversal committed:
Changed context `no_mod` updating to allow overriding.

*Title

AllUniversal committed:
Added Cloud Nine implementation (untested)

+Title

AllUniversal committed:
Added LSP def for new `get_ranks.flags` default value

+Title

! Cloud Nine seems to be working! No lag on my end.

Aurelius7309 commented:
Re: Cloud Nine: Card:update() does not return automatically when a method is present on the center. I'm not sure why, but if there isn't a mod that relies on adding update functionality to vanilla jokers with the original still being run (which I doubt, and it would be bad practice anyway), that behaviour can easily be changed. Otherwise I suggest a patch to disable the relevant part of the vanilla function. As is, your fix should work but won't remove the original overhead that is no longer needed.

AllUniversal committed:
Updated lovely patch for calling a card's center's update function

*It now returns, shouldn't cause problems hopefully

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Added `SMODS.lowest_and_highest_rank()` + LSP defs

+Title

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

nh6574 committed:
Merge branch 'main' into pr/AllUniversal/838

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Added `SMODS.match_cards()` and related functions

+Title, these functions allow for checking which cards of a list of cards match certain conditions, like having a certain enhancement and rank.
-> The goal with this is to ultimately allow easy custom hand detection. Currently I'd still have to implement a bipartite matching function or something to efficiently de-duplicate card matches.

*Changed `Card:is_ranks()` to take a map instead of a list as its `ranks` param.

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Added `SMODS.get_hand_from_matching()` to allow using matching for deduplicated poker hand detection.

+Title, this function uses recursive bipartite matching! :)
-> This allows completely arbitrary poker hand detection, like a poker hand that requires two lucky cards and a red seal, or similar.
- Removed unnecessary map of matcher conditions.

AllUniversal committed:
LSP defs

+/*Title

AllUniversal committed:
More LSP defs + `matcher` condition `check_function`

+/*Title, the `check_function` condition allows for arbitrary matching of cards.

AllUniversal committed:
Another `matcher` condition; `suit`

+Title, dunno how I forgot that one lmao

AllUniversal committed:
Moved `matcher` code to separate PR.

-/*Title

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

Aurelius7309 commented on src/utils.lua:
```diff
@@ -315,35 +315,33 @@ function SMODS.modify_rank(card, amount, manual_sprites)
             elseif behavior.random then
                 rank_key = pseudorandom_element(
-                    rank_data.next,
+                    SMODS.get_keys(rank_data.next),
```
This creates inconsistent RNG due to `pairs` having undefined order

AllUniversal replied:
Would sorting the keys alphabetically resolve this?

Aurelius7309 replied:
Assuming usage is restricted to string keys only, yes.

Aurelius7309 commented on src/utils.lua:
```diff
@@ -1931,13 +1955,58 @@ function SMODS.update_context_flags(context, flags)
     if flags.denominator then context.denominator = flags.denominator end
     if flags.cards_to_draw then context.amount = flags.cards_to_draw end
     if flags.saved then context.game_over = false end
+    if flags.ranks_fixed or flags.ranks_changed then
```
This construct is awkward to use. Using keys over rank objects seems like the better choice, given it spares you from having to type out `SMODS.Ranks` all the time when you don't care about any properties of the rank object. It also seems more usable to have a single return key for changed ranks (this can just be `modify`) and a boolean flag like `fixed = true` if you want to override existing ranks

AllUniversal replied:
I've changed `Card:get_ranks()` to now use a single `ranks` return field and the `fixed` flag as you suggested it.
Additionally it now only takes rank keys as return values.
Edit: Oh and I opted for `ranks` over `modify` for clarity + I'm currently looking at a general quantum API where this would probably be necessary.

Aurelius7309 commented on src/overrides.lua:
```diff
@@ -619,74 +619,424 @@ G.FUNCS.change_stake = function(args)
+		return nil
+	end
+
+	if next(valid_pairs) then
```
This merge results in duplicate cards in the merged hand if cards have more than one rank that each belongs to a different pair. I'm unsure if this causes any issues, but deduplicating this is simple and the safer choice

Aurelius7309 commented on src/utils.lua:
```diff
@@ -2992,6 +3084,176 @@ function SMODS.is_eternal(card, trigger)
+
+function Card:get_ranks(flags) -- Returns a map of "SMODS.Rank"s, sanitized to ONLY be "SMODS.Rank"s -> Rank keys or rank ids are converted to SMODS.Rank 
+    if not self.playing_card then return {} end
+    local default_ranks = (SMODS.has_no_rank(self) and {}) or {[SMODS.Ranks[self.base.value]] = true}
```
`no_rank` is weak, causing `SMODS.has_no_rank` to yield inconsistent results if any quantum rank effects adds a rank to an otherwise rankless card. Conversely, `SMODS.has_no_rank` doesn't reliably indicate if a rank is "effectively rankless". These interactions make the function quite unintuitive to use; perhaps it's better to differentiate between general no-rank checks and ones used in checking ranks?

AllUniversal replied:
Hmm, so should I move the current checks from `SMODS.has_no_rank()` over into `Card:get_ranks()` to decide the `default_ranks`, and then have `SMODS.has_no_rank()` use `Card:get_ranks()` to check whether a card has no ranks?

Aurelius7309 replied:
I guess that makes sense. The alternative is an additional argument for `SMODS.has_no_ranks` — not sure right now what's cleaner

AllUniversal replied:
With some recent thinking for a potential general quantum API, I realized there's a fundamental problem with having `SMODS.has_no_rank()` inside `Card:get_ranks()` to define the default values; 
`SMODS.has_no_rank()` currently supports quantum *enhancements*, in a general quantum API, those would *also* use a respective `SMODS.has_no_enhancement()` function in their getter function to define default values, but that causes a loop; 
1. The getter is called, 
2. has_no is called to define default values, 
3. has_no calls all/some getters for *other* quantum fields (e.g. seals, editions, enhancements, etc.), 
-> A loop is entered.

I'm currently thinking about how to approach a solution; It might be unavoidable to have `no_rank` and the like *not* support quantum enhancements/etc, thus e.g. a QE Stone card would *not* count as `no_rank`/`no_suit`.
Still looking into a better solution. (My brain has taken up a side-gig as an industrial oven.)

AllUniversal replied:
I have now opened a new Draft PR #1351, it tackles this is issue by not letting QEs etc. affect `no_...`/`any_...` of a card. In theory there may still be a way to have it work, but I dunno. :)

Aurelius7309 commented on src/utils.lua:
```diff
@@ -2992,6 +3084,176 @@ function SMODS.is_eternal(card, trigger)
+    return #cards > 0
+end
+
+function SMODS.lowest_and_highest_rank(cards)
```
What is this meant to be used for? Sort order is not generally indicative of one rank being higher/lower than another

AllUniversal replied:
It's meant for getting the highest or lowest card, for effects like Raised Fist, but you're right to say that `sort_nominal` isn't really the best measure of "highest" or "lowest" with ranks. Any suggestions on what to replace it with? Might have to add a field to `SMODS.Rank`.
Edit: That said, I do believe `sort_nominal` is also what's used to sort the cards in hand, so there's some logic to using it here.

Aurelius7309 replied:
The issue with any approach is that a generalized set of ranks doesn't need to have a total ordering, so highest/lowest ranks as a concept may not make sense. In that case `sort_nominal` probably does make the most sense

AllUniversal replied:
I agree, guess I'll leave it as is then.

AllUniversal committed:
First changes according to review.

*/-Title, includes refactors to `get_ranks()`

AllUniversal committed:
Further refactoring

*Title

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal committed:
Merge remote-tracking branch 'upstream/main' into Quantum-Ranks

AllUniversal commented:
I've now merged this PR into #1351! I will now be working on this and all the new stuff over there, so I'm going to close this PR.
